### PR TITLE
JDK-8243204 add class load time checking of abstract super classes of…

### PR DESCRIPTION
**Pull request by hseigel**
Please review this chagne.  It adds checks of super classes of inline classes.  It checks that an inline class's super class is  either java.lang.Object or an abstract class.  And, if abstract, checks that the super, and its super, etc. do not have instance fields, nor synchronized instance methods, or non-empty constructors, nor constructors with args.

The fix was tested with tiers 1 - 3.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * Frederic Parain ([fparain](@fparain) - Committer)

### Download
`$ git fetch https://git.openjdk.java.net/valhalla pull/34/head:pull/34`
`$ git checkout pull/34`

### CODE REVIEW

**Comment by fparain** [comment]
Shouldn't `ClassFileParser::check_super_of_inline_type()` also check that no super classes implement the `java.lang.IdentityObject` interface?

**Comment by hseigel** [comment]
The update to the pull request implements John's "bit A and bit B" idea posted in the java-valhalla slack channel.  Instead of A and B, the flags are named invalid_inline_super and invalid_identity_super.  In this pull request, invalid_identity_super exists but is never set nor read.  That is for a future change.

Flag invalid_inline_super is set to TRUE in a class's InstanceKlass, if, at class load time, the class is determined to be an invalid super class for an inline type.

Checking that the class's constructor is abstract will be done at a future date once there is compiler support.

File: src/hotspot/share/classfile/classFileParser.cpp [hunk 1]
```diff
@@ -6874,6 +6898,38 @@ void ClassFileParser::parse_stream(const ClassFileStream* const stream,
...
+      } else if (method->name() == vmSymbols::object_initializer_name()) {
+        if (method->signature() != vmSymbols::void_method_signature() ||
+            !method->is_vanilla_constructor()) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
```
Comment (fparain) [inline src/hotspot/share/classfile/classFileParser.cpp:6931]: It seems this method could be made static.

File: src/hotspot/share/classfile/classFileParser.cpp [hunk 2]
```diff
@@ -6874,6 +6898,38 @@ void ClassFileParser::parse_stream(const ClassFileStream* const stream,
   // all bytes in stream read and parsed
 }
 
+// Return true if the specified class is not a valid super class for an inline type.
+// A valid super class for an inline type is abstract, has no instance fields,
+// does not implement interface java.lang.IdentityObject (checked elsewhere), has
+// an empty body-less no-arg constructor, and no synchronized instance methods.
```
Comment (fparain) [inline src/hotspot/share/classfile/classFileParser.cpp:6904]: The comment should be more explicit about the fact that it is performing a local test, without considering super types. The final decision taking account super types being performed in `fill_instance_klass()`.

**Comment by hseigel** [comment]
Today's updated change improves the comment for function is_invalid_super_for_inline_type() and makes it a static function.  Also, it moves the InstanceKlass invalid_inline_super and invalid_identity_super flags into misc_flags.

File: src/hotspot/share/classfile/classFileParser.cpp (hunk 2 above)
Comment (hseigel) [reply src/hotspot/share/classfile/classFileParser.cpp]: The latest change updates this comment.

File: src/hotspot/share/classfile/classFileParser.cpp (hunk 1 above)
Comment (hseigel) [reply src/hotspot/share/classfile/classFileParser.cpp]: Done.

File: src/hotspot/share/oops/instanceKlass.hpp [hunk 3]
```diff
@@ -258,6 +258,9 @@ class InstanceKlass: public Klass {
   // _misc_flags.
   bool            _is_marked_dependent;  // used for marking during flushing and deoptimization
 
+  bool _invalid_inline_super;   // if true, invalid super type for an inline type.
+  bool _invalid_identity_super; // if true, invalid super type for an identity type.
```
[the remark this answers is not in the document]
Comment (hseigel) [reply src/hotspot/share/oops/instanceKlass.hpp]: Done.  Thanks for the suggestion!

**Pull request closed without merging**